tests/index.test.ts

```diff
@@ -55,4 +56,200 @@ describe('index', () => {
+    const eP = new ErrorPropagate<string>(undefined, { cause: 'string' });
+    const eF = new ErrorFixed(undefined, { cause: new SyntaxError() });
+    const eD = new ErrorDefault<number>(undefined, { cause: 123 });
+    expect(eP.cause).toBe('string');
+    expect(eF.cause).toBeInstanceOf(SyntaxError);
+    expect(eD.cause).toBe(123);
+  });
+  test('JSON encoding/decoding', () => {
+    const e = new AbstractError();
+    expect(AbstractError.fromJSON(e.toJSON())).toBeInstanceOf(AbstractError);
+    expect(AbstractError.fromJSON(e.toJSON()).toJSON()).toStrictEqual(
+      e.toJSON(),
+    );
+    const eJSON = {
+      type: 'AbstractError',
+      data: {
+        message: 'some message',
+        timestamp: '2022-05-07T09:16:06.632Z',
+        data: {},
+        cause: undefined,
+      },
+    };
+    const e2 = AbstractError.fromJSON(eJSON);
+    expect(e2).toBeInstanceOf(AbstractError);
+    expect(e2.message).toBe(eJSON.data.message);
+    expect(e2.timestamp).toStrictEqual(new Date(eJSON.data.timestamp));
+    expect(e2.data).toStrictEqual(eJSON.data.data);
+    expect(e2.cause).toStrictEqual(eJSON.data.cause);
+  });
+  describe('JSON serialiation and deserialisation', () => {
+    // Demonstrates an extended error with its own encoding and decoding
+    class SpecificError<T> extends AbstractError<T> {
+      public static fromJSON<T extends Class<any>>(
+        this: T,
+        json: any,
+      ): InstanceType<T> {
+        if (
+          typeof json !== 'object' ||
+          json.type !== this.name ||
+          typeof json.data !== 'object' ||
+          typeof json.data.message !== 'string' ||
+          typeof json.data.num !== 'number' ||
+          ('stack' in json.data && typeof json.data.stack !== 'string')
+        ) {
+          throw new TypeError(`Cannot decode JSON to ${this.name}`);
+        }
+        const e = new this(json.data.num, json.data.message);
+        e.stack = json.data.stack;
+        return e;
+      }
+      public num: number;
+      public constructor(num: number, message: string) {
+        super(message);
+        this.num = num;
+      }
+      public toJSON() {
+        const obj = super.toJSON();
+        obj.data.num = this.num;
+        return obj;
+      }
+    }
+    class UnknownError<T> extends AbstractError<T> {}
+    // AbstractError classes, these should be part of our application stack
+    const customErrors = {
+      AbstractError,
+      SpecificError,
+      UnknownError,
+    };
+    // Standard JS errors, these do not have fromJSON routines
+    const standardErrors = {
+      Error,
+      TypeError,
+      SyntaxError,
+      ReferenceError,
+      EvalError,
+      RangeError,
+      URIError,
+    };
+    function replacer(key: string, value: any): any {
+      if (value instanceof Error) {
+        return {
+          type: value.constructor.name,
+          data: {
+            message: value.message,
+            stack: value.stack,
+          },
```

Comment:
Node v16 should also have the `cause` property too.

Reply:
Node v16 is mostly supporting 2021, and while it does support the `cause `property, the `cause` propert is actually ES2022. So we don't really support it yet.

Reply:
See: https://node.green/